oswaldo ordaz madrazo 1/10/23

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

In [8]:
import numpy as np
import scipy.optimize as opt

# Coeficientes de costo por gramo de cada ingrediente
c = np.array([0.013,
              0.008,
              0.010,
              0.002,
              0.005,
              0.001])

# Coeficientes de restricciones de nutrientes
Aub = np.array([
    [-0.10, -0.20, -0.15, 0.00, -0.04, 0.00],  # Proteína
    [-0.08, -0.10, -0.11, -0.01, -0.01, 0.00],  # Grasa
    [0.001, 0.005, 0.005, 0.10, 0.15, 0.00],  # Fibra
    [0.002, 0.005, 0.007, 0.002, 0.008, 0.00]   # Sal
])

# Valores objetivo de las restricciones de nutrientes
Bub = np.array([-8, -6, 2, 0.4])

sol = opt.linprog(c, Aub, Bub)
sol

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.48
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   0.000e+00]
            nit: 1
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
                 marginals: [ 6.600e-03  0.000e+00  1.200e-03  1.200e-03
                              4.200e-03  1.000e-03]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 4.000e+00  0.000e+00  1.700e+00  1.000e-01]
                 marginals: [-0.000e+00 -8.000e-02 -0.000e+00 -0.000e+00]
 mip_node_count: 0
 mip_

se ultiliza pura carne para minimizar

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

In [17]:

# Coeficientes de costo de distribución por millón de KW
c = np.array([5, 2, 7, 3, 6, 6, 6, 1, 2])

# Coeficientes de restricciones de oferta de plantas
A_ub = np.array([
    [1, 1, 1, 0, 0, 0, 0, 0, 0],  # Planta 1
    [0, 0, 0, 1, 1, 1, 0, 0, 0],  # Planta 2
    [0, 0, 0, 0, 0, 0, 1, 1, 1]   # Planta 3
])

# Valores objetivo de las restricciones de oferta
B_ub = np.array([80, 40, 60])

# Coeficientes de restricciones de demanda de ciudades
A_eq = np.array([
    [1, 0, 0, 1, 0, 0, 1, 0, 0],  # Guadalajara
    [0, 1, 0, 0, 1, 0, 0, 1, 0],  # León
    [0, 0, 1, 0, 0, 1, 0, 0, 1]   # Morelia
])

# Valores objetivo de las restricciones de demanda
B_eq = np.array([70, 40, 70])

sol = opt.linprog(c, A_ub, B_ub, A_eq, B_eq)
sol

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

el costo total es de 540

# planta 1
- se mandan 30 mkw a gdl
- se mandan 40 mkw a leon
- se mandan 10 mkw a morelia

# planta 3
- manda 60 mkw a morelia 